### ISSUE TITLE
Printing verse ranges fails to include a paragraph style

### ISSUE THREAD

**Issue by mnjames** [sillsdev/ptx2pdf#812]
If I have a text like this:
```
\s انسانی حقوق کا عالمی منشور 
\p 
\v 1 اقوامِ متحدہ کی جنرل اسمبلی نے ۱۰؍ دسمبر ؁۱۹۴۸ء کو ”انسانی حقوق کا عالمی منشور“ منظور کر کے اس کا اعلانِ عام کیا۔ 
\m
\v 2 اگلے صفحات پر اس منشور کا مکمل متن درج ہے۔ اس تاریخی کارنامے کے بعد اسمبلی نے اپنے تمام ممبر ممالک پر زور دیا کہ وہ بھی اپنے اپنے ہاں اس کا اعلانِ عام کریں اور اس کی نشر و اشاعت میں حصہ لیں۔
\v 3 مثلاً یہ کہ اسے نمایاں مقامات پر آویزاں کیا جائے۔ اور خاص طور پر اسکولوں اور تعلیمی اداروں میں اسے پڑھ کر سنایا جائے اور اس کی تفصیلات واضح کی جائیں، اور اس ضمن میں کسی ملک یا علاقے کی سیاسی حیثیت کے لحاظ سے کوئی امتیاز نہ برتا جائے۔
```
and want to print just "GEN 12:1", the pdf comes out looking very bad. This is because the .sfm which PTXprint creates looks like this:
```
⋮
\toc1 پیدایٔش
\v 1 اقوامِ متحدہ کی جنرل اسمبلی نے ۱۰؍ دسمبر ؁۱۹۴۸ء کو ”انسانی حقوق کا عالمی منشور“ منظور کر کے اس کا اعلانِ عام کیا۔
\p
```

In other words, when told to print verse 1 PTXprint just blindly starts at the `\\v 1` marker and doesn't look before it for context. I think this is probably a problem everywhere, but isn't as noticeable in LTR languages because when printing text that isn't proceeded by a paragraph style, LTR languages don't fail spectacularly. But in RTL languages it's obvious because the document ends up being left justified, has the verse numbers on the left, has the indent on the left, etc.

-----------

At the very least, if the range being printed is immediately preceded by some sort of paragraph marker, I'd argue that that marker should be included in the text to be printed.

-----------

Other situations seem a little less clear-cut, but I still think PT should be able to handle them. If I tell it to print "GEN 12:3", it's a bit more ambiguous what should be done, and I could suggest a few options which might all be considered solutions:
1) PTXprint, realizing that if it just starts printing from `\\v 3` that the USFM it creates won't be well-formed, blindly inserts the "default" paragraph marker of `\\p`.
2) PTXprint reads backwards until it find the whatever paragraph style verse 3 is part of (`\\m` here).
3) PTXprint continues to use badly formed USFM, but in RTL languages if the text isn't inside a paragraph style, the text still follows the RTL direction and is right justified, indented, etc.
4) Option 1 or 2 is implemented _only if_ the "Auto-Correct USFM" checkbox is selected.

**Comment by viverechristus** [comment]
I'm getting the same problem. If I print only ACT 3:1-10, it comes out like this:

![image](https://user-images.githubusercontent.com/915182/213072312-52cce267-3f93-43bd-826d-b04efdd67b1f.png)

The SFM output is missing the \c marker, and maybe other stuff:

```
\id ACT ‑ Jarai Cambodia Khmer Script
\toc3 ឞ្វៈម៍
\toc1 ឞ្វៈម៍ ង៉ៈ
\toc2 ឞ្វៈម៍ ង៉ៈ
\h ឞ្វៈម៍ ង៉ៈ
\mt1 ឞ្វៈម៍ ង៉ៈ
\v 1 ហ្មវ ហា រើយ ពេត្រូស ឋ្វា ញូ យ៉ូហាន ឌីគ ណាវ ពើ សាង យ៉ា
```

Printing an entire chapter (say, ACT 3) is fine, no problems.

**Comment by davidg-sil** [comment]
The question of context is really tricky, and the simple fact is that people will have different ideas, probably depending on what their goal is.  Consider this:
```
\c 11
\v 1 ....
\v 26 ....
\mt  The Account of Abraham
\mr 11:27-25:11
\s Abraham's family
\p
\v 27 ...
```
What counts as 'context' for verse 27?

1. Just  paragraph mark?
2. Paragraph mark and the \s heading?
3. Paragraph mark, \s, \mt, \mr section headings
4. The chapter number too?

For an 'account of Abraham' trial publication, I'd want to argue that the correct answer is 4. Others might say 3.
And if it's just 3 verses, I'd be quite surprised to see the `\mt` and `\mr` in there.

**Comment by mnjames** [comment]
I agree that it's tricky and that different people will have different assumptions. For #‌2 I'd like the \s heading to appear, and for #‌4, I'd like the chapter number to appear. But I'm very, very flexible on that and wouldn't be upset if the team went a different direction.

The one that I'm very strongly arguing for is #‌1. I think that if the verse is _immediately_ preceded by a paragraph mark (of any kind), it should be included in the printout. Otherwise you're trying to create a pdf using badly formed USFM. 

(And if the verse isn't _immediately_ preceded by a paragraph mark, PTXprint either ought to find the correct one by reading backward of just throw in a `\p`.)

**Comment by viverechristus** [comment]
I'd agree with @mnjames that it matters less what exactly the markup is and more that it's simply well-formed. Our own use case is preparing a few verse for testing or reading a portion with our translation team for review. If we wanted to prepare isolated verses for special publishing purposes (e.g., a list of memory verses prepared for distribution), we'd use LibreOffice, not PTXprint.

**Comment by markpenny** [comment]
@mnjames There was a fix for #894 which might have fixed this issue as well. Could you please check-n-confirm, and close the issue if all is well now. (Note that the auto-inserted \zsetref code might cause issues when exporting USFM to SAB, so keep your eyes open for surprises there.)

**Comment by mnjames** [comment]
No, adding that \zsetref milestone doesn't cause any change in my output.

My .sfm output file is very basic:
```
\id GEN ‑ XYZ
\h Genesis
\toc3 Genesis
\toc2 Genesis
\toc1 Genesis
\mt1 Genesis
\zsetref|bkid="GEN" chapter="11" verse="1"\*\v 1 First sentence of text.
```

It looks like there still isn't a \p or other paragraph marker before \v 1, so it's just continuing with the \mt1 formatting (and not starting a new line).

**Comment by davidg-sil** [comment]
Bother. @mhosken  Obviously more work to do here :(
